tests/php/Dev/DeprecationTest.php

```diff
@@ -8,103 +8,33 @@
-    {
-        Deprecation::notification_version('1.0.0');
-        $this->assertNull(Deprecation::notice('2.0', 'Deprecation test failed'));
-    }
-
-    public function testEqualVersionTriggersNotice()
+    public function testNotice()
     {
+        $message = implode(' ', [
+            'SilverStripe\Dev\Tests\DeprecationTest->testNotice is deprecated.',
+            'My message.',
+            'Called from PHPUnit\Framework\TestCase->runTest.'
+        ]);
         $this->expectError();
-        Deprecation::notification_version('2.0.0');
-        Deprecation::notice('2.0.0', 'Deprecation test passed');
-    }
-
-    public function testBetaVersionDoesntTriggerNoticeWhenDeprecationDoesntSpecifyReleasenum()
-    {
-        Deprecation::notification_version('2.0.0-beta1');
-        $this->assertNull(Deprecation::notice('2.0', 'Deprecation test failed'));
-        $this->assertNull(Deprecation::notice('2.0.0', 'Deprecation test failed'));
-    }
-
-    public function testGreaterVersionTriggersNotice()
-    {
-        $this->expectError();
-        Deprecation::notification_version('3.0.0');
-        Deprecation::notice('2.0', 'Deprecation test passed');
-    }
-
-    public function testNonMatchingModuleNotifcationVersionDoesntAffectNotice()
-    {
-        Deprecation::notification_version('1.0.0');
-        Deprecation::notification_version('3.0.0', 'my-unrelated-module');
-        $this->callThatOriginatesFromFramework();
-    }
-
-    public function testMatchingModuleNotifcationVersionAffectsNotice()
-    {
-        $this->expectError();
-        Deprecation::notification_version('1.0.0');
-        Deprecation::notification_version('3.0.0', 'silverstripe/framework');
-        $this->callThatOriginatesFromFramework();
-    }
-
-    public function testMethodNameCalculation()
-    {
-        $this->assertEquals(
-            TestDeprecation::get_method(),
-            static::class . '->testMethodNameCalculation'
-        );
-    }
-
-    public function testScopeMethod()
-    {
-        $this->expectError();
-        $this->expectErrorMessage('DeprecationTest->testScopeMethod is deprecated. Method scope');
-        Deprecation::notification_version('2.0.0');
-        Deprecation::notice('2.0.0', 'Method scope', Deprecation::SCOPE_METHOD);
-    }
-
-    public function testScopeClass()
-    {
-        $this->expectError();
-        $this->expectErrorMessage('DeprecationTest is deprecated. Class scope');
-        Deprecation::notification_version('2.0.0');
-        Deprecation::notice('2.0.0', 'Class scope', Deprecation::SCOPE_CLASS);
-    }
-
-    public function testScopeGlobal()
-    {
-        $this->expectError();
-        $this->expectErrorMessage('Global scope');
-        Deprecation::notification_version('2.0.0');
-        Deprecation::notice('2.0.0', 'Global scope', Deprecation::SCOPE_GLOBAL);
+        $this->expectErrorMessage($message);
+        Deprecation::$notice_level = E_USER_NOTICE;
```

Comment:
Why swap to `E_USER_NOTICE`? You can use `$this->expectDeprecation()` and `$this->expectDeprecationMessage()` instead of the error variants https://phpunit.readthedocs.io/en/9.5/writing-tests-for-phpunit.html#writing-tests-for-phpunit-exceptions-examples-errortest-php

Reply:
`expectedDeprecated()` does not work with `E_USER_DEPRECATED` https://github.com/laminas/laminas-di/pull/30#issuecomment-927585210

Reply:
It does work - that's what that method is explicitly for. The issue you linked to merely points out that there was a change at some stage which changed the default config for converting deprecation warnings into exceptions. Previously it did the conversion by default, but now we have to set some configuration to make it do so.

There are three ways we can do this:
1. Add `convertDeprecationsToExceptions="true"` to the `<phpunit> tag in `phpunit.xml.dist`. This will need to be done for the framework module itself, as well as anything like installer and recipes which also run framework's unit tests. It will result in ALL deprecation warnings being converted to exceptions, which we probably don't want.
    ```xml
    <phpunit bootstrap="tests/bootstrap.php" colors="true" convertDeprecationsToExceptions="true">
    ```
1. Add the following method to the `DeprecationTest` class, which will enable converting deprecations to exceptions for the duration of the test - but will also keep converting them _after_ this test. There's no way to turn that off unfortunately.
     ```php
     public function run(?TestResult $result = null): TestResult
     {
         if ($result === null) {
             $result = $this->createResult();
         }
         $result->convertDeprecationsToExceptions(true);
         return parent::run($result);
     }
     ```
1. Add a custom error handler for the `DeprecationTest` in `setUp()` and remove it in `tearDown()`. This has the advantage of explicitly only converting deprecation notices that are fired during the execution of tests in this class. Note that this must not be done in `setUpBeforeClass()` or it will be registered before phpunit's own error handler and cause other problems.
    ```php
    // This is phpunit's error handler
    private $oldErrorHandler;

    protected function setUp(): void
    {
        parent::setUp();
        $this->oldErrorHandler = set_error_handler(function (int $errorNumber, string $errorString, string $errorFile, int $errorLine) {
            if ($errorNumber === E_USER_DEPRECATED) {
                throw new Deprecated($errorString, $errorNumber, $errorFile, $errorLine);
            }
            // Make sure to call phpunit's error handler if it's not a deprecation notice
            return $this->$oldErrorHandler($errorNumber, $errorString, $errorFile, $errorLine);
        });
    }
    
    protected function tearDown(): void
    {
        Deprecation::$notice_level = E_USER_DEPRECATED;
        Deprecation::disable();
        restore_error_handler();
        parent::tearDown();
    }
    ```
    
All that said, while *I* think this is worth doing both to remove the unnecesary ability to change the notice type and to make sure we're actually testing the deprecation notice being thrown - I don't feel strongly enough to put my foot down. So if you want to leave it as is just say so and I'll mark this conversation resolved.

Reply:
The most "correct" way to do this would be option 1. `convertDeprecationsToExceptions="true"`, however this could easily have side effects, such as non-silverstripe dependencies throwing deprecation warnings. I'm not sure what effect this would have with custom Silverstripe error handlers

Option 2 is too risky

Option3 is the most viable, though could easily have side effects messing with other custom error handlers that are set within Silverstripe e.g. graphql

I think it's safer to just stick with crude yet functional approach we already have